syntax = "proto3";

package dfs_service;

service DFSService {

    // Add your service calls here

    // 1. REQUIRED (Parts 1 & 2): A method to store files on the server
    rpc Store (stream FileChunk) returns (StoreResponse);

    // 2. REQUIRED (Parts 1 & 2): A method to fetch files from the server
    rpc Fetch (FetchRequest) returns (stream FileChunk);

    // 3. REQUIRED (Parts 1 & 2): A method to list all files on the server
    rpc List (ListRequest) returns (ListResponse);

    // 4. REQUIRED (Parts 1 & 2): A method to get the status of a file on the server

    // 5a. REQUIRED (Part 2 only): A method to request a write lock from the server
    rpc AcquireWriteLock(LockRequest) returns (LockResponse);

    // 5b. REQUIRED (Part 2 only): A method to release a write lock from the server
    rpc ReleaseWriteLock(LockRequest) returns (LockResponse);

    // 6. REQUIRED (Part 2 only): A method named CallbackList to handle asynchronous file listing requests
    //                            from a client. This method should return a listing of files along with their
    //                            attribute information. The expected attribute information should include name,
    //                            size, modified time, and creation time.
    rpc CallbackList(CallbackListRequest) returns (CallbackListResponse);

    // 7. REQUIRED (Part 2 only): A method to delete a file from the server
    rpc Delete (DeleteRequest) returns (DeleteResponse);

    // 8. Any other methods you deem necessary to complete the tasks of this assignment


}

// Add your message types here
message FileChunk {
    string client_id = 1;
    string file_name = 2;
    bytes data = 3;
    uint32 crc = 4;
    int64 mtime = 5;
}

message StoreResponse {
    bool ok = 1;
    string message = 2;
}

message FetchRequest {
    string client_id = 1;
    string file_name = 2;
    uint32 crc = 4;
    int64 mtime = 5;
}

message DeleteRequest {
    string client_id = 1;
    string file_name = 2;
}

message DeleteResponse {
    bool ok = 1;
    string message = 2;
}

message ListRequest {
    string client_id = 1;
}

message ListResponse {
    map<string, int64> file_info=1; // filename -> unix epoch millis
}

message LockRequest {
    string client_id = 1;
    string file_name = 2;
}

message LockResponse {
    bool granted = 1;
    string holder = 2;
}

message FileInfo {
    string file_name = 1;
    uint32 crc = 2;
    int64 mtime = 3;
    bool deleted = 4;
}

message CallbackListRequest {
    string client_id = 1;
    string name = 2;
}

message CallbackListResponse {
    repeated FileInfo files = 1;
}